pennylane/devices/qutrit_mixed/sampling.py

```diff
@@ -0,0 +1,460 @@
+    """
+    # apply diagonalizing gates
+    state = _apply_diagonalizing_gates(mps, state, is_state_batched)
+
+    total_indices = get_num_wires(state, is_state_batched)
+    wires = qml.wires.Wires(range(total_indices))
+
+    def _process_single_shot(samples):
+        processed = []
+        for mp in mps:
+            if isinstance(mp, SampleMP):
+                res = math.squeeze(_process_samples(mp, samples, wires))
+            elif isinstance(mp, CountsMP):
+                res = _process_counts_samples(mp, samples, wires)
+            elif isinstance(mp, ExpectationMP):
+                res = _process_expval_samples(mp, samples, wires)
+            elif isinstance(mp, VarianceMP):
+                res = _process_variance_samples(mp, samples, wires)
+            else:
+                raise NotImplementedError
+            processed.append(res)
+        return tuple(processed)
+
+    # if there is a shot vector, build a list containing results for each shot entry
+    if shots.has_partitioned_shots:
+        processed_samples = []
+        for s in shots:
+            # Like default.qubit currently calling sample_state for each shot entry,
+            # but it may be better to call sample_state just once with total_shots,
+            # then use the shot_range keyword argument
+            samples = sample_state(
+                state,
+                shots=s,
+                is_state_batched=is_state_batched,
+                wires=wires,
+                rng=rng,
+                prng_key=prng_key,
+            )
+            processed_samples.append(_process_single_shot(samples))
+
+        return tuple(zip(*processed_samples))
+
+    samples = sample_state(
+        state,
+        shots=shots.total_shots,
+        is_state_batched=is_state_batched,
+        wires=wires,
+        rng=rng,
+        prng_key=prng_key,
+    )
+
+    return _process_single_shot(samples)
+
+
+def _measure_hamiltonian_with_samples(
+    mp: List[SampleMeasurement],
+    state: np.ndarray,
+    shots: Shots,
+    is_state_batched: bool = False,
+    rng=None,
+    prng_key=None,
+):
+    """Sums expectation values of Sum Observables"""
+    # the list contains only one element based on how we group measurements
+    mp = mp[0]
+
+    # if the measurement process involves a Hamiltonian, measure each
+    # of the terms separately and sum
+    def _sum_for_single_shot(s):
+        results = measure_with_samples(
+            [ExpectationMP(t) for t in mp.obs.terms()[1]],
+            state,
+            s,
+            is_state_batched=is_state_batched,
+            rng=rng,
+            prng_key=prng_key,
+        )
+        return sum(c * res for c, res in zip(mp.obs.terms()[0], results))
+
+    unsqueezed_results = tuple(_sum_for_single_shot(type(shots)(s)) for s in shots)
+    return [unsqueezed_results] if shots.has_partitioned_shots else [unsqueezed_results[0]]
+
+
+def _measure_sum_with_samples(
+    mp: List[SampleMeasurement],
```

Comment:
Added to make expectation values of Sum Observables work since it wouldn't be difficult to add

Reply:
the difference is mostly just the s_product, you are right it makes more sense to make them one function.